feat: Separating Submission of Salary Slips and Creation of Journal Entry in Payroll Entry

viralkansodiya opened the pull request:
#818  
Hey @ruchamahabal  , please look at this issue 

I have developed this feature , can you please check and Merge


https://github.com/frappe/hrms/assets/141210323/bb864f9b-81c3-4e6f-9208-de3b399fa06f

saurabh6790 committed:
chore: merge common code and break larger functions into smaller functions

saurabh6790 committed:
feat: cache holiday list

saurabh6790 committed:
feat: cache leave type mapper

saurabh6790 committed:
feat: cache putput of get_value api for salary component

saurabh6790 committed:
feat: add missing indexes

saurabh6790 committed:
feat: rewrite get_lwp_ppl for date

saurabh6790 committed:
feat: do not run postprosess if calling from salary slip

saurabh6790 committed:
feat: add composite index

saurabh6790 committed:
feat: add caching

saurabh6790 committed:
feat: add property setter for joining_date and relieving_date, set ssa in object to reuse

saurabh6790 committed:
feat: add indexing

saurabh6790 committed:
fix: caching

saurabh6790 committed:
feat: use cahced version of salary structure in doc mapping

saurabh6790 committed:
fix: use get_cached_doc instead of get_value with cache true

saurabh6790 committed:
fix: get employee holiday dates

saurabh6790 committed:
fix: do not use cached version in testing

saurabh6790 committed:
fix: dates based on joining and relieving date

saurabh6790 committed:
fix: test case and caching

ruchamahabal committed:
ci: run parallel tests

saurabh6790 committed:
chore: multiple fixes

saurabh6790 committed:
chore: resolve merge conflicts

saurabh6790 committed:
fix: test case

ruchamahabal committed:
test(fix): create missing fixtures

ruchamahabal committed:
Merge pull request #785 from ruchamahabal/fix-payroll-entry-reload

ruchamahabal committed:
fix: perm errors during expense claim creation for ESS users

ruchamahabal committed:
fix(Payroll JE): link employee in advance return entries

krantheman committed:
fix(ux): '(optional)' text in field labels

krantheman committed:
refactor: add checkboxes for selecting employees

krantheman committed:
fix: remove unused Allocate button

ruchamahabal committed:
Merge pull request #794 from ruchamahabal/fix-ess-access-for-claims

krantheman committed:
feat: add Select fields for Dates and Allocation Based On

krantheman committed:
refactor: add DataTable instead of MultiCheck

krantheman committed:
feat: add back end functionality for Dates Based On

krantheman committed:
feat: add Leave Policy Assignment functionality

krantheman committed:
feat: add Dates Based On Joining Date

krantheman committed:
fix: whack Default option for Select fields

krantheman committed:
feat: reset fields after Allocating Leave

krantheman committed:
fix: failed list scope

krantheman committed:
refactor: single frm.set_value() call

ruchamahabal committed:
fix: link advance against deduction JV + minor refactor

krantheman committed:
fix(Shift Type): rename misleading Enable Entry/Exit Grace Period checkbox label (#784)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

niraj2477 committed:
fix: Allow float for `Maximum Leave Allocation Allowed` (#789)

ruchamahabal committed:
chore: fix employee name visibility & adjust col widths

ruchamahabal committed:
fix: move shift start & end time columns to the start

ruchamahabal committed:
chore(formatting): separate logically related blocks by blank lines

ruchamahabal committed:
fix: respect system datetime format for shift timing columns

ruchamahabal committed:
feat: show attendance ID column to quickly navigate to record

ruchamahabal committed:
fix(report query): apply groupby after where clause

ruchamahabal committed:
chore: add report link to Shift workspace

ruchamahabal committed:
test: fix date formatting

ruchamahabal committed:
test: advance deduction in accrual journal entry

krantheman committed:
refactor: use Duration instead of Time for Late Entry/Early Exit By

krantheman committed:
test: add basic tests

krantheman committed:
fix(ux): move Employment Type to right column

krantheman committed:
feat: add Report to Shift & Attendance Workspace

krantheman committed:
fix(linter): no newline at end of file

ruchamahabal committed:
fix: add permission for employee

mergify[bot] committed:
Merge pull request #775 from krantheman/shift_attendance_report

feat: add Shift Attendance Report

krantheman committed:
fix(Shift Type): rename `enable_(entry/exit)_grace_period` to `enable_(late_entry/early_exit)_marking` (#803)

Renames fieldname for previously renamed label.

See: [784](https://github.com/frappe/hrms/pull/784)

krantheman committed:
fix(Shift Attendance): rename enable_grace_period fields as enable_marking (#813)

ruchamahabal committed:
Merge pull request #796 from ruchamahabal/fix-emp-adv-return

ruchamahabal committed:
fix: allow optional tax component addition & removal (#814)

niyazrazak committed:
fix(FnF): don't fetch outstanding statements on every save (#797)

The user cannot leave the receivables tables empty in the doctype. With each save, loans and employee advances are added to the table.

saurabh6790 committed:
fix: do not exclude process_salary_structure

ruchamahabal committed:
feat: add Maximum Encashable Leaves setting in Leave Type + minor refactor

ruchamahabal committed:
fix: set salary structure if not found

ruchamahabal committed:
feat: allow overwriting encashable days

ruchamahabal committed:
fix: use existing holiday list fn instead of changing other fn defs

ruchamahabal committed:
refactor: replace db cache with `get_cached_value` for fetching >1 values

ruchamahabal committed:
fix: incorrect db cache call

ruchamahabal committed:
fix: reduce db/fn calls + fix hasattr check

ruchamahabal committed:
fix: add missing employee mapping from structure -> slip

ruchamahabal committed:
fix: `get_single_value` calls are cached by default

ruchamahabal committed:
fix: lwp calculation, missing filter in query & typo

ruchamahabal committed:
refactor: replace db cache with `get_cached_value` for fetching >1 values

ruchamahabal committed:
fix: holiday date type casting

ruchamahabal committed:
test: add regression test for half day lwp in leave-based payroll

ruchamahabal committed:
refactor: `calculate_lwp_or_ppl_based_on_leave_application`

- remove dead code

ruchamahabal committed:
refactor(caching): replace unnecessary hashes with value cache

ruchamahabal committed:
fix: add missing cache invalidation

ruchamahabal committed:
test(fix): clear cache in tests

ruchamahabal committed:
perf: cache payroll period fn calls

ruchamahabal committed:
fix: create constants for cache keys

ruchamahabal committed:
Merge pull request #470 from saurabh6790/performance-fixes-salary-slip

perf: faster Payroll :- Part 2 (salary slip)

ruchamahabal committed:
revert: enable variable based on taxable salary for Income Tax component (#820)

mergify[bot] committed:
fix: Create Salary Slips button not visible on failure (backport #826) (#827)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

ruchamahabal committed:
refactor: backward compatible caching (#833)

ruthra-kumar committed:
fix: unwanted gain/loss on base currency exp claims

ruthra-kumar committed:
test: invalid gain/loss JE for Expense Claim

ruchamahabal committed:
Merge pull request #848 from ruthra-kumar/gain_loss_issue_in_expense_claim

viralkansodiya commented:
Hey @ruchamahabal any update of this PR?
if you need any help from my side Please tell me,
we are always ready to contribute our skills

anandbaburajan committed:
fix: update repay_from_salary and payroll_payable_account fields (#851)

deepeshgarg007 committed:
fix: Employee-wise loan repayment processing

deepeshgarg007 committed:
chore: linting issues

deepeshgarg007 committed:
chore: linting issues

deepeshgarg007 committed:
chore: linting issues

deepeshgarg007 committed:
chore: check is lending app installed

deepeshgarg007 committed:
Merge pull request #855 from deepeshgarg007/employe_wise_loan_repayment

fix: Employee-wise loan repayment processing

zeel-codder committed:
fix: Order skills in Interview Feedback Doctype same as skills of Interview Round DocType (#847)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
refactor: move Select Employees section below Set Leave Details

ruchamahabal committed:
chore: reduce containers for parallel tests

ruchamahabal committed:
fix: set setup wizard company to the one used in tests

ruchamahabal committed:
fix(test): set default holiday list before tests

ruchamahabal committed:
chore: remove unnecessary create_company calls

ruchamahabal committed:
Merge pull request #599 from ruchamahabal/ci-parallel-tests

ruchamahabal committed:
ci: number of builds for coverage

ruchamahabal committed:
fix: company fetching & filter in Salary Structure Assignment (#870)

ruchamahabal committed:
test: create missing fixtures

ruchamahabal committed:
ci: fix upload coverage data

ruchamahabal committed:
test(fix): add customer as a dependency for project profitability

ruchamahabal committed:
Merge pull request #873 from ruchamahabal/ci-fix

ruchamahabal committed:
fix(Payroll Entry): show Create Salary Slips button only after saving (#874)

krantheman committed:
feat: Select Employees filters those employees for whom leaves have already been allocated

krantheman committed:
refactor: check field instead of select for Allocation Based On and Custom Range instead of empty option for Dates Based On

krantheman committed:
feat: show Select Employees Section only after reuqired fields have been filled out

krantheman committed:
refactor(ui): move Filters Section in between

ruchamahabal committed:
fix(UX): make performance feedback dialog minimizable (#877)

ruchamahabal committed:
fix(UX): make feedback dialogs static (#879)

zeel-codder committed:
feat: Enable Allow Import (via Data Import Tool) for DocTypes Job Applicant and Job Opening (#882)

niraj2477 committed:
fix: Add close button in goal view

krantheman committed:
fix: amended Expense Claims' resetting of sanctioned amounts (#885)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
fix(Salary Slip): Income Tax Breakup not using set currency (#887)

anandbaburajan committed:
fix: fetch logic for repay_from_salary in loan_repayment [dev] (#890)

ankush committed:
fix: alternate formula eval implementation

Current frappe.safe_eval transforms code so if you have nested
iterations with too much depth then it can hit recursion limit of python.

There's no workaround for this and people need large formulas in some
countries, so this is alternate implementation for that.

ruchamahabal committed:
fix: use single underscore fn & fix assertions in tests

ankush committed:
Merge pull request #902 from ankush/eval

fix: alternate formula eval implementation

ruchamahabal committed:
fix: compensatory leave request for Half Day & WFH (#906)

mergify[bot] committed:
fix: required_apps (backport #911) (#913)

Co-authored-by: Francisco Roldán <franciscoproldan@gmail.com>

ruchamahabal committed:
feat(Payroll Settings): Consider Marked Attendance on Holidays

ruchamahabal committed:
test: refactor employee creation and usage

ruchamahabal committed:
test: consider marked attendance on holidays

ruchamahabal committed:
feat(UX): show payroll settings in payment days tab for context

ruchamahabal committed:
Merge pull request #915 from ruchamahabal/payroll-attendance

ruchamahabal committed:
style: add bottom padding to employee section

ruchamahabal committed:
fix: don't reset leave details on refresh

ruchamahabal committed:
fix: carry forward field description

ruchamahabal committed:
feat: allow editing encashment amount in salary structure after submit

ruchamahabal committed:
fix: inform the user about overwriting encashable days

ruchamahabal committed:
fix: filter section collapsing automatically

ruchamahabal committed:
fix: set better defaults & reduce db calls

ruchamahabal committed:
refactor: set leave period dates for context

ruchamahabal committed:
refactor: better error handling and notifying

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -136,9 +136,29 @@ frappe.ui.form.on('Payroll Entry', {
 			frm.add_custom_button(__("Create Salary Slips"), function() {
 				frm.trigger("create_salary_slips");
 			}).addClass("btn-primary");
+		} else if (
+			(!frm.doc.journal_entry_submitted
+			&& frm.doc.salary_slips_submitted)
+			|| (frm.doc.__onload && frm.doc.__onload.submitted_ss)
+		) {
+			frm.events.add_journal_entry_button(frm);
```
You will have to remove accrual entry creation from `submit_salary_slips_for_employee`. It is currently created automatically on salary slip submission

Also remove accrual entry info from this message shown on submission since now its a separate action.

![image](https://github.com/frappe/hrms/assets/24353136/f05e18ad-1d12-4ed5-adec-9ab9e1982d92)

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -136,9 +136,29 @@ frappe.ui.form.on('Payroll Entry', {
+	add_journal_entry_button: function(frm){
+		frm.add_custom_button(__("Create Journal Entry"), function () {
+			frappe.call({
+				doc:frm.doc,
+				method: 'make_accrual_jv_entry',
```
```suggestion
				doc: frm.doc,
				method: "make_accrual_jv_entry",
```

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -136,9 +136,29 @@ frappe.ui.form.on('Payroll Entry', {
 	},
 
+	add_journal_entry_button: function(frm){
+		frm.add_custom_button(__("Create Journal Entry"), function () {
```
```suggestion
		frm.add_custom_button(__("Make Accrual Entry"), function () {
```
Can we call this button "Make Accrual Entry" to be inline with "Make Bank Entry"?
Accordingly change the function names, labels and messages on client-side
Also, make it primary

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.json:
```diff
@@ -324,12 +325,21 @@
-  }
+  },
+  {
+   "default": "0",
+   "fieldname": "journal_entry_submitted",
+   "fieldtype": "Check",
+   "hidden": 1,
+   "label": "Journal Entry Submitted",
+   "read_only": 1,
+   "no_copy": 1
```
On second thought, do we need this field at all? We are anyway checking if there are any existing JV accounts on load. We have `salary_slips_created` and `salary_slips_submitted` flags because these operations are queued for slips > 30. But in this case, JVs are created right away.

ruchamahabal commented:
Also, modify testcases once you finish the changes

viralkansodiya commented:
Hey @ruchamahabal Please review a changes I did all the changes as per your suggestion

krantheman committed:
feat: add Advanced Filters section

niraj2477 committed:
feat: Add status dropdown button

niraj2477 committed:
fix: Update list view & json schema

niraj2477 committed:
fix: Update set_status function

niraj2477 committed:
Merge branch 'develop' into close-goal

krantheman committed:
feat: add Advanced Filters

krantheman committed:
refactor: redirect Bulk Leave Policy Assignment to Leave Control Panel

krantheman committed:
feat: add Leave Control Panel to Leaves workspace

ruchamahabal committed:
refactor: add `actual_start_date` & `actual_end_date` properties based on DOJ & Relieving Dates

ruchamahabal committed:
fix(payroll): incorrect absent days if relieving date is set for a latter month

anandbaburajan committed:
refactor: rename loan type to loan product

anandbaburajan committed:
refactor: rename loan type to loan product

ruchamahabal committed:
test: remove assertion for amount in consider marked attendance on holidays test (#930)

ruchamahabal committed:
fix: status button groups & formatting

ruchamahabal committed:
Merge pull request #886 from niraj2477/close-goal

ruchamahabal committed:
fix: shift fetch fails if shift's end date & actual end date are different

ruchamahabal committed:
refactor: smaller functions for getting shift details

ruchamahabal committed:
fix: absent marking for missing attendance dates fails for an older assignment date

niraj2477 committed:
fix(Payroll): Change "Preview Salary Slips" dropdown to autocomplete (#935)

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

ruchamahabal committed:
test: mark absent for dates with no attendance for an older date

ruchamahabal committed:
test: mark absent for dates with no attendance for midnight shift

ruchamahabal committed:
test(refactor): absent records for dates with no attendance

barredterra committed:
refactor: migrate translations from ERPNext

barredterra committed:
refactor: migrate translations to ERPNext

ruchamahabal committed:
fix: Overflow error for date operation when day changes

- simplify conditions

ruchamahabal committed:
fix: shift fetch fails in before shift margin if shift's start & actual start date are different

ruchamahabal committed:
fix: handle false positive cases for shift fetching in shift boundaries

ruchamahabal committed:
test: false positives cases for shift fetching in shift margins

ruchamahabal committed:
refactor: smaller functions for flagging shifts outside assignment period

+ add comments

ruchamahabal committed:
chore: add comments for fetching shifts for a date to consider margins

ruchamahabal committed:
fix: incorrect attendance date for different start & actual start dates

ruchamahabal committed:
fix: absent marking fails if shift's end date & actual end date are different

ruchamahabal committed:
perf: use generator fn for date range to find prev/next assignment

ruchamahabal committed:
refactor: return sorted list of attendance dates for predictable behaviour

ruchamahabal committed:
Merge pull request #937 from ruchamahabal/fix-attendance

ruchamahabal committed:
chore: remove unused file `employee_loan.js`

ruchamahabal committed:
Merge pull request #928 from anandbaburajan/loan_product_rename

ruchamahabal committed:
Merge pull request #942 from barredterra/erpnext-translations

niraj2477 committed:
fix: Exclude  employee from report

ruchamahabal committed:
fix(Expense Claim): respect system precision for rounding

ruchamahabal committed:
test: rounding in claims

ruchamahabal committed:
fix: replace hard-coded route passed to `set_route`

- remove unused functions & fix message titles

ruchamahabal committed:
Merge pull request #954 from ruchamahabal/fix-rounding-in-claims

krantheman committed:
feat(Employee): update job applicant status on insert

ruchamahabal committed:
fix: show datatable empty state instead of hiding the section

add comments for advanced filters working

ruchamahabal committed:
perf: replace for loop with join query

- add type hints

ruchamahabal committed:
test: allocation based on joining date & filters

ruchamahabal committed:
Merge pull request #809 from krantheman/leave_control_panel

krantheman committed:
feat(Employee): update job offer status and submit

ruchamahabal committed:
fix: fetching dates based on leave period

anandbaburajan committed:
test: add product_code in create_loan_product (#955)

niraj2477 committed:
fix: Update status condition

krantheman committed:
feat(Employee Onboarding): add Mark as Completed button

krantheman committed:
fix: save after updating

ruchamahabal committed:
ci: use mariadb instead of mysql (#963)

krantheman committed:
test: add test for Mark as Completed

krantheman committed:
test: perf

krantheman committed:
chore: move create_job_applicant to test_utils to prevent circular import

krantheman committed:
test: add test for update_job_applicant_and_offer

ruchamahabal committed:
perf: faster `get_employees` for previewing salary slips

ruchamahabal committed:
fix: enqueue payroll entry cancellation for salary slips > 50

ruchamahabal committed:
fix(realtime): restrict updates to current user only

ruchamahabal committed:
test: payroll entry cancellation queueing

ruchamahabal committed:
Merge pull request #964 from ruchamahabal/perf-payroll-process

krantheman committed:
fix: display of unsubmitted Leave Policies and inactive Leave Periods (#969)

krantheman committed:
fix: permission/mandatory errors

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

krantheman committed:
feat: add msgprint for changing Job Applicant/Offer status

mergify[bot] committed:
Merge pull request #953 from niraj2477/gh-738

fix(Report): Employees working on a holiday

ruchamahabal committed:
chore: fix formatting

ruchamahabal committed:
Merge pull request #961 from krantheman/feat-employee-onboarding

ruchamahabal committed:
refactor: make auto-update message more informative

- place job applicant update/validate functions together

- replace implicit submission for job offer with a message

ruchamahabal committed:
Merge pull request #957 from krantheman/feat-job-applicant

ruchamahabal committed:
fix(Payroll): fetching error message from message log (#981)

ruchamahabal committed:
fix: add bottom padding for Import Attendance section

ruchamahabal committed:
style(fix): link icon & new feedback button border not visible in perf feedback

ruchamahabal committed:
style(Org Chart): replace box-shadow with borders & fix font styles

ruchamahabal committed:
fix: node card avatar

ruchamahabal committed:
style: change node & connector colors from blue -> gray

ruchamahabal committed:
style(fix): edit button

ruchamahabal committed:
chore: replace `cur_frm` usage

ruchamahabal committed:
Merge pull request #985 from ruchamahabal/espresso-fixes

ruchamahabal committed:
Merge branch 'develop' into fix-payroll-payment-days

ruchamahabal committed:
Merge pull request #925 from ruchamahabal/fix-payroll-payment-days

ruchamahabal committed:
refactor: rename Encashment Threshold Days field to Non Encashable Leaves

ruchamahabal committed:
test: max encashable leaves & non-encashable leaves settings

ruchamahabal committed:
chore: enable non-negative check for encashment settings

niraj2477 committed:
feat: Add sender in payroll setting & HR Setting

ruchamahabal committed:
feat: add Actual Encashable Days field to show no. of leaves eligible for encashment

ruchamahabal committed:
fix: set leave allocation value in db on submit

ruchamahabal committed:
patch: only update empty leave encashment fields

ruchamahabal committed:
Merge pull request #817 from ruchamahabal/enhance-leave-encashment

ruchamahabal committed:
chore: add new frappe hr logo as app icon

ruchamahabal committed:
chore: update new logo in readme

ruchamahabal committed:
ci: update ci config to consider v15

ruchamahabal committed:
chore: update base version for develop

ruchamahabal committed:
chore: fix logo in readme

ruchamahabal committed:
chore: update readme

creative-paramu committed:
feat: Add Select/Unselect All options in Mark Attendance Dialog (#1005)

* feat:Add Unselect all in Mark Attendance Dialog

* chore: fix formatting

---------

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

anandbaburajan committed:
fix: close employee loan on write off (#991)

niraj2477 committed:
feat: Add employee grade filter in payroll entry (#960)

ruchamahabal committed:
refactor: use fetch from and `get_sender_email` function

ruchamahabal committed:
Merge branch 'develop' into gh-707

ruchamahabal committed:
Merge pull request #988 from niraj2477/gh-707

feat: Add sender in Payroll Settings & HR Settings

ruchamahabal committed:
perf: batch and commit auto attendance processing

ruchamahabal committed:
test(fix): setup shift type for monthly attendance sheet tests

ruchamahabal committed:
fix: bump employee chunk size

ruchamahabal committed:
Merge pull request #1022 from ruchamahabal/batch-attendance

perf: batch and commit auto attendance processing

ruchamahabal committed:
feat(util): add provision to get holidays except weekly offs from holiday list (#1011)

zuron7 committed:
chore: Added default permission for HR Manager to submit F&F Statement (#1021)

* Added default permission for HR Manager to submit F&F Statement

* chore: fix incorrect json

* chore: update modified timestamp

---------

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

viralkansodiya commented:
Raising new PR
